File: src/main/scala/uk/gov/ons/methods/MarkingBetweenLimits.scala
package uk.gov.ons.methods

import org.apache.spark.sql.DataFrame

class MarkingBetweenLimits(val dfIn: DataFrame) extends BaseMethod {
  if (dfIn == null) throw new Exception("DataFrame cannot be null")

  import uk.gov.ons.methods.impl.MarkingBetweenLimitsImpl._

  /** Scala function that calls version 1 of the marking between limits method.
    *
    * @author kayleigh.bellis@ext.ons.gov.uk
    * @version 1.0
    * @param input            DataFrame - The input DataFrame for the function to be applied to
    * @param valueColumnName  String - Name of the Column that will be divided by its previous
    * @param upperLimit       Any - The upper bound of what the ratio can be to be flagged 1
    * @param lowerLimit       Any - The lower bound of what the ratio can be to be flagged 1
    * @param partitionColumns List[String] - A list of column names that will be partition on when getting the
    *                         previous value
    * @param orderColumns     List[String] - A list of column names that will be order by when getting the
    *                         previous value
    * @param newColumnName    String - Name of the new Column
    * @return DataFrame
    */
  def markingBetweenLimitsMethod(input: DataFrame = dfIn, valueColumnName: String, upperLimit: Any, lowerLimit: Any,
                              partitionColumns: Seq[String], orderColumns: Seq[String],
                              newColumnName: String): DataFrame = {
    mandatoryArgCheck(valueColumnName, upperLimit, lowerLimit, newColumnName,
      Seq(partitionColumns, orderColumns).flatten)
    input.markingBetweenLimitsMethod(valueColumnName, upperLimit, lowerLimit,
      partitionColumns, orderColumns, newColumnName)
  }
}

object MarkingBetweenLimits {
  def markingBetweenLimits(df: DataFrame): MarkingBetweenLimits = new MarkingBetweenLimits(df)
}

File: python/sml/methods/markingBetweenLimits.py
from pyspark.sql import DataFrame


class MarkingBetweenLimits():
    def __init__(self, df=None):
        if df is None: raise TypeError

        self._df = df
        self._jmbl = self._df._sc._jvm.uk.gov.ons.api.java.methods.MarkingBetweenLimitsFactory.markingBetweenLimits(
            self._df._jdf)

    def __mandatoryArgumentCheck(self, *args):
        for arg in args:
            if arg is None:
                raise TypeError

    def markingBetweenLimits(self, df=None, valueColumnName=None, upperLimit=None, lowerLimit=None,
                             partitionColumns=None, orderColumns=None, newColumnName=None):
        """
        A Python wrapper for the markingBetweenLimits function

        :param df: a DataFrame
        :param valueColumnName: Name of the value column to mark
        :param upperLimit: The upper limit to compare to
        :param lowerLimit: The lower limit to compare to
        :param partitionColumns: The columns in which to partition on when getting the previous value
        :param orderColumns: The columns in which to order by when getting the previous value
        :param newColumnName: Name of the new Column to make
        :return: pyspark.sql.DataFrame
        """
        self.__mandatoryArgumentCheck(valueColumnName, upperLimit, lowerLimit, partitionColumns, orderColumns,
                                      newColumnName)
        if df is None: df = self._df
        return DataFrame(
            self._jmbl.markingBetweenLimits(df._jdf, valueColumnName, upperLimit, lowerLimit, partitionColumns,
                                            orderColumns, newColumnName), df.sql_ctx)


def markingBetweenLimits(df):
    return MarkingBetweenLimits(df)
